src/appleseed/renderer/modeling/object/curveobjectreader.cpp

```diff
@@ -75,369 +79,268 @@ namespace renderer
-            split_and_store(object.ref(), curve, split_count);
         }
-    }
-
-    input.close();
-
-    if (input.bad())
-    {
-        RENDERER_LOG_ERROR("failed to load curve file %s: i/o error.", filepath.c_str());
-        return object;
-    }
-
-    stopwatch.measure();
 
-    RENDERER_LOG_INFO(
-        "loaded curve file %s (%s curve%s) in %s.",
-        filepath.c_str(),
-        pretty_uint(curve_count).c_str(),
-        curve_count > 1 ? "s" : "",
-        pretty_time(stopwatch.get_seconds()).c_str());
+        void push_curve3(int stride)
+        {
+            for (int32 i = 0; i < m_vertices.size() - 3; i = i + stride) {
+                GVector3 points[4] = {m_vertices[i], m_vertices[i + 1], m_vertices[i + 2], m_vertices[i + 3]};
+                GScalar widths[4] = {m_widths[i], m_widths[i + 1], m_widths[i + 2], m_widths[i + 3]};
+                GScalar opacities[4] = {m_opacities[i], m_opacities[i + 1], m_opacities[i + 2], m_opacities[i + 3]};
+                GColor3 colors[4] = {m_colors[i], m_colors[i + 1], m_colors[i + 2], m_colors[i + 3]};
 
-    return object;
+                split_and_store(Curve3Type(points, widths, opacities, colors), m_split_count);
+            }
+        }
+    };
 }
 
-auto_release_ptr<CurveObject> CurveObjectReader::load_mitsuba_curve_file(
+
+auto_release_ptr<CurveObject> CurveObjectReader::read(
     const SearchPaths&      search_paths,
     const char*             name,
     const ParamArray&       params)
 {
-    // todo: fix for big endian CPUs.
-
-    auto_release_ptr<CurveObject> object(CurveObjectFactory().create(name, params));
-
-    const string filepath = to_string(search_paths.qualify(params.get("filepath")));
+    const string filepath = params.get<string>("filepath");
 
-    FILE* file = fopen(filepath.c_str(), "rb");
-    if (file == nullptr)
+    const size_t basis = params.get_optional<size_t>("basis", 2);
+    if (basis < 1 && basis > 4)
     {
-        RENDERER_LOG_ERROR("failed to open curve file %s.", filepath.c_str());
-        return object;
+        RENDERER_LOG_ERROR("curves basis choice must be between 1 and 4 but was " FMT_SIZE_T ".", basis);
+        return auto_release_ptr<CurveObject>(nullptr);
     }
 
-    char signature[12];
-
-    if (fread(signature, 1, 11, file) != 11)
-       RENDERER_LOG_ERROR("failed to load curve file %s: unknown signature.", filepath.c_str());
-
-    signature[11] = '\0';
-
-    if (strcmp(signature, "BINARY_HAIR") != 0)
+    const float radius = params.get_optional<float>("radius", 0.01f);
+    if (radius <= 0.0f)
     {
-        fclose(file);
-        RENDERER_LOG_ERROR("failed to load curve file %s: unknown signature.", filepath.c_str());
-        return object;
+        RENDERER_LOG_ERROR("curves radius must be greater than zero.");
+        return auto_release_ptr<CurveObject>(nullptr);
     }
 
+    GenericCurveFileReader reader(search_paths.qualify(filepath.c_str()).c_str(), radius, basis);
```

Comment:
Maybe move `search_paths.qualify(filepath.c_str()).c_str()` to separate variable to improve readability.

Are you sure that you need to call `c_str()` on `filepath`?